### ISSUE TITLE
TechDraw - Auxiliary View

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#5810]
Issue imported from https://tracker.freecad.org/view.php?id=3609

* **Reporter:** normandc
* **Date submitted:** 9/24/2018
* **FreeCAD version:** 
* **Category:** Feature
* **Status:** new
* **Tags:** 

# Original report text

Per Wikipedia:
[quote=Wikipedia]An auxiliary view is an orthographic view that is projected into any plane other than one of the six primary views. These views are typically used when an object contains some sort of inclined plane. Using the auxiliary view allows for that inclined plane (and any other significant features) to be projected in their true size and shape. The true size and shape of any feature in an engineering drawing can only be known when the Line of Sight (LOS) is perpendicular to the plane being referenced.[/quote]
https://en.wikipedia.org/wiki/Engineering_drawing#Auxiliary_views

The auxiliary view has a "viewing plane indicator" (really no clue how it's called) which usually is a dotted line with arrows, similar to a section view. It may depend on drawing standards. The auxiliary view is usally linked and aligned to its parent view, although it may be possible to break the alignment if there is not enough space on the page besides the parent view.

A possible work flow would be to select a linear edge on a view then launch the auxiliary view tool. A FreeCAD file with TechDraw page is attached, with a pseudo auxiliary view made with the standard NewView tool. Attached svg was edited to manually add the plane indicator with view label.

# Other bug information

* **Priority:** normal
* **Severity:** minor
* **Category:** Feature
* **Updated:** 2/6/2021

**Comment by WandererFan** [comment]
![FreeCAD-TechDraw-AuxiliaryView-request](https://user-images.githubusercontent.com/3587161/191385393-a5387192-3298-46bc-adf8-0e9134c64a5d.svg)
[TechDraw_Feature_Request_Auxiliary_View_Example_nc1.fcstd.zip](https://github.com/FreeCAD/FreeCAD/files/9611927/TechDraw_Feature_Request_Auxiliary_View_Example_nc1.fcstd.zip)

**Comment by WandererFan** [comment]
Additional information: 
https://wiki.freecadweb.org/TechDraw_SectionExamples#Auxiliary_view
https://www.forum.freecadweb.org/viewtopic.php?p=646546#p646546

**Comment by WandererFan** [comment]
Description of the Work

The DocumentObject DrawAuxiliaryView inherits DrawViewPart, with some added properties:

- BaseView - the DrawViewPart that this DAV relates to
- AuxiliaryDirection  - projection direction relative to BaseView. Used to calculate the absolute projection direction and rotation.
- ReferenceLabel - the text that appears in both the BaseView and the DAV.

A new graphics object (ex. QGIViewAuxiliary) is not required as the painting of the view is the same as a regular view.

QGIViewPart will need to be modified to draw the reference Text and arrow on top of the BaseView.

A graphics object for the Reference and Arrow will probably be needed.

The existing ViewProvider for DVP can likely be used, but need to confirm that changes to AuxiliaryDirection and ReferenceLabel can be detected in ViewProviderViewPart. If necessary, a light weight view provider can be derived as is done for sections.

A new command will need to be added to TechDraw/Gui/Command.cpp and TechDraw/Gui/Workbench.cpp.

A new task dialog will be needed to set direction. 

A new wiki page will be needed. 

Refer to the checklist for incidental tasks.
[newFeatureChecklist.txt](https://github.com/FreeCAD/FreeCAD/files/10198691/newFeatureChecklist.txt)

**Comment by pierreporte** [comment]
A **150 € reward** will be given to the contributor that writes a pull request, solving this issue, that gets merged. It will be paid by the FreeCAD Project Association. It is to be considered an incentive. Grant applications are welcome, but the total amount will exclude the reward.

**Comment by FC-FBXL5** [comment]
> A new task dialog will be needed to set direction (see TechDraw/Gui/TaskSectionView.cpp).

Please reconsider. The direction should be set graphically by selecting vertices or straight edges in the base view. This resembles the workflow in descriptive geometry done manually that has been used for ages. Other CAD apps don't do it this way without reason. (this applies to section views, too)

A task dialog should only have a checkbox to reverse the direction.

**Comment by FC-FBXL5** [comment]
I'm looking forward to an auxiliary view feature integrated in TD and prepared a preliminary wiki page containing a proof of concept macro in Python.
I chose this way as I'm still unable to compile C++ code. Maybe this helps someone in the meantime while waiting for the one to pick the bounty. See https://wiki.freecad.org/TechDraw_AuxiliaryView

**Comment by FC-FBXL5** [comment]
It has been a while since I have created my first workbench called Templater (available via the Addon Manager) that also includes a tool called AuxView, see https://wiki.freecad.org/Templater_AuxView
It uses  two selected points or one selected straight edge to define the auxiliary view's orientation and has a simple task panel to set the marker, toggle the orientation between along or across the given line, and toggle between forward and reverse direction.
It has no dashed line as a "viewing plane indicator" as it It wasn't necessary according to the standards I was used to, and it only has a single arrow to show the viewing direction of the auxiliary view in the base view.

Too bad that I'm still unable to translate this into C++ ...

**Comment by meaqua9420** [comment]
I'm looking into implementing this.

Would a selection-first MVP be acceptable for the initial PR?

Scope I have in mind:
- add `DrawAuxiliaryView` derived from `DrawViewPart`, with `BaseView`, a local auxiliary/reference direction, and `ReferenceLabel`
- create it from one selected straight edge or two selected vertices in the base `DrawViewPart`
- derive `Direction` / `XDirection` from the base view
- draw a single reference arrow + label on the base view
- keep the task panel small: marker label, along/across orientation, and forward/reverse direction

More advanced marker styling or standards-specific options could be follow-up work. Does this match the intended direction?

**Comment by pierreporte** [comment]
@meaqua9420 Yes you can do more than one PR to implement the full thing. Please note that the bounty would be paid only when the whole issue is resolved, not just for a MVP.

**Comment by FC-FBXL5** [comment]
@WandererFan 
I think it would be an easy task for you to implement this auxiliary view since it is a section view without cutting anything of the displayed object. What do you think?
Needed inputs: 
- Base view z-axis: **Direction** property - check!
- Base view x-axis: **XDirection** property - check! 
- The angle of the XDirection from the horizontal page axis: **Rotation** property - check!
- 2 points or one straight line to obtain the angle of the AuxViews x-axis measured from the horizontal page axis. At this point we can add/subtract 180^ to reverse the view direction, or add/subtract 90^ to toggle between normal to or along the selected (connection) line. This angle will become the **Rotation** property of the AuxView.

Calculate 2D:
- Subtract the first mentioned angle from the second one to obtain the angle between the x-axes of Base view and AuxView in the Base view which is also the angle in 3D.

Calculate 3D:
- Use the angle of the step before to rotate the **XDirection** of the Base view around its **Direction** and so obtain the 3D vector to use as the AuxView's **XDirection** property.
- The y-axis of the AuxView equals the **Direction** property of the Base view so that the cross-product of AuxView's **XDirection** and Base view's **Direction** give the **Direction** (the z-axis) of the AuxView.

Finally we have the **Direction**, **XDirection**, and **Rotation** properties of the AuxView.
Adding a simple view and setting these properties would complete the creation of an AuxView, that seems to be easier than creating a section view.
In contrast to a section view there is no need to acquire further information from the 3D space, any input can be found in the Base view or can be calculated.

Both section ans auxiliary view would benefit from being able to project the global origin and from being able to measure distances and create dimensions from/to this projection, sections could be defined in the Base view instead of the 3d view. This projection could also help to align AuxView and Base view. - But that would be another issue.

**Comment by franklincg** [comment]
Hi, I'm interested in working on this bounty. Could you confirm the current reward amount and whether the issue is still available for a new contributor to work on?
